lib/syscall_shim/uk_prsyscall.c

```diff
@@ -0,0 +1,709 @@
+{
+	UK_ASSERT(sb);
+	UK_ASSERT(uk_streambuf_buf(sb));
+
+	if (PTRISERR((void *) retval) && (retval != 0)) {
+		uk_streambuf_shcc(sb, fmtf, ERROR);
+		uk_streambuf_printf(sb, "%s (%ld)", strerror((int) -retval),
+				    retval);
+		uk_streambuf_shcc(sb, fmtf, RESET);
+	} else if (type == PT_BOOL) {
+		/* Replace boolean with OK (here, errors are already catched) */
+		uk_streambuf_shcc(sb, fmtf, OK);
+		uk_streambuf_strcpy(sb, "OK");
+		uk_streambuf_shcc(sb, fmtf, RESET);
+	} else {
+		/* Use same formatting as for parameters */
+		pr_param(sb, fmtf, type, retval);
+	}
+}
+
+/*
+ * Helper macros for pretty print a system call request (PR_SYSCALL,
+ * VPR_SYSCALL) and to pretty print the return code of a system call
+ * (PR_SYSRET).
+ */
+#define __PR_SYSCALL_PARAM_PRINT(i, sb, type, arg)			\
+	do {								\
+		if ((i) != 0)						\
+			uk_streambuf_strcpy((sb), ", ");		\
+		pr_param((sb), fmtf, (type), (long) (arg));		\
+	} while (0)
+
+#define PR_SYSCALL(sb, fmtf, syscall_num, ...)				\
+	do {								\
+		uk_streambuf_shcc(sb, fmtf, SYSCALL);			\
+		uk_streambuf_strcpy(sb, uk_syscall_name((syscall_num))); \
+		uk_streambuf_shcc(sb, fmtf, RESET);			\
+		uk_streambuf_strcpy(sb, "(");				\
+		VARG2_EXPAND(__PR_SYSCALL_PARAM_PRINT, sb, __VA_ARGS__); \
+		uk_streambuf_strcpy(sb, ")");				\
+	} while (0)
+
+#define __VPR_SYSCALL_PARAM_DECL(i, unused, type)			\
+	long UK_CONCAT(arg, i);
+#define __VPR_SYSCALL_PARAM_VALOAD(i, ap, type)				\
+	UK_CONCAT(arg, i) = va_arg(ap, long);
+#define __VPR_SYSCALL_PARAM_PRINT(i, sb, type)				\
+	if ((i) != 0)							\
+		uk_streambuf_strcpy((sb), ", ");			\
+	pr_param((sb), fmtf, (type), UK_CONCAT(arg, i));		\
+
+#define VPR_SYSCALL(sb, fmtf, syscall_num, ap, ...)			\
+	do {								\
+		VARG_EXPAND(__VPR_SYSCALL_PARAM_DECL, 0, __VA_ARGS__)	\
+		VARG_EXPAND(__VPR_SYSCALL_PARAM_VALOAD, ap, __VA_ARGS__) \
+									\
+		uk_streambuf_shcc(sb, fmtf, SYSCALL);		\
+		uk_streambuf_strcpy(sb, uk_syscall_name((syscall_num))); \
+		uk_streambuf_shcc(sb, fmtf, RESET);			\
+		uk_streambuf_strcpy(sb, "(");				\
+		VARG_EXPAND(__VPR_SYSCALL_PARAM_PRINT, sb, __VA_ARGS__) \
+		uk_streambuf_strcpy(sb, ")");				\
+	} while (0)
+
+#define PR_SYSRET(sb, fmtf, type, rc)					\
+	do {								\
+		uk_streambuf_strcpy(sb, " = ");				\
+		pr_retcode((sb), (fmtf), (type), (rc));			\
+	} while (0)
+
+/*
+ * Pretty print a system call request to a stream buffer
+ */
+static void pr_syscall(struct uk_streambuf *sb, int fmtf,
+		       long syscall_num, long rc, va_list args)
+{
+	switch (syscall_num) {
+#ifdef HAVE_uk_syscall_brk
+	case SYS_brk: {
+			long addr = (long) va_arg(args, long);
+
+			PR_SYSCALL(sb, fmtf, syscall_num,
+					PT_VADDR, addr);
+			PR_SYSRET(sb, fmtf, PT_VADDR, rc);
+		}
+		break;
+#endif /* HAVE_uk_syscall_brk */
+
+#ifdef HAVE_uk_syscall_open
+	case SYS_open:
+		VPR_SYSCALL(sb, fmtf, syscall_num, args, PT_PATH, PT_OFLAGS);
+		PR_SYSRET(sb, fmtf, PT_FD, rc);
+		break;
+#endif /* HAVE_uk_syscall_openat */
```

Comment:
In the comment "open" should be instead of "openat".

Reply:
:+1: